### PR TITLE
Feature #3131 schema tags

### PR DESCRIPTION
**Pull request by marcobiedermann**
I've added some schema markup tags for:
- Breadcrumb
- Calendar
- Events
- Gallery
- Image
- Navigation
- News
- Page
- Pagination

Source: http://schema.org/

### CODE REVIEW

**Comment by Toflar** [comment]
+1 for those!

**Comment by leofeyer** [comment]
@marcobiedermann Deckt dieser PR auch #5968 ab?

**Comment by marcobiedermann** [comment]
@leofeyer Sure, https://github.com/marcobiedermann/core/commit/fe5f4d99176140fb75c48c8271af2cb7f3965648 fixes #5968

File: system/modules/calendar/templates/calendar/cal_default.html5 [hunk 1]
```diff
@@ -19,8 +19,9 @@
         <td class="<?php echo $day['class']; ?>">
           <div class="header"><?php echo $day['label']; ?></div>
           <?php foreach ($day['events'] as $event): ?>
-            <div class="event cal_<?php echo $event['parent']; ?><?php echo $event['class']; ?>">
-              <a href="<?php echo $event['href']; ?>" title="<?php echo $event['title']; ?> (<?php if ($event['day']): echo $event['day']; ?>, <?php endif; echo $event['date']; if ($event['time']): ?>, <?php echo $event['time']; endif; ?>)"<?php echo $event['target']; ?>><?php echo $event['link']; ?></a>
+            <div class="event cal_<?php echo $event['parent']; ?><?php echo $event['class']; ?>" itemscope itemtype="http://schema.org/Event">
+              <a href="<?php echo $event['href']; ?>" title="<?php echo $event['title']; ?> (<?php if ($event['day']): echo $event['day']; ?>, <?php endif; echo $event['date']; if ($event['time']): ?>, <?php echo $event['time']; endif; ?>)"<?php echo $event['target']; ?> itemprop="url"><span itemprop="name"><?php echo $event['link']; ?></span></a>
```
Comment (leofeyer) [inline system/modules/calendar/templates/calendar/cal_default.html5]: Why did you add `<span itemprop="name">` around the URL?

File: system/modules/calendar/templates/calendar/cal_default.html5 [hunk 2]
```diff
@@ -19,8 +19,9 @@
...
+              <time datetime="<?php echo $this->datetime; ?>" class="date invisible" itemprop="startDate" content="<?php echo $this->datetime; ?>"><?php echo $event['date']; ?><?php if ($event['time']): ?>, <?php echo $event['time']; endif; ?></time>
```
Comment (leofeyer) [inline system/modules/calendar/templates/calendar/cal_default.html5]: Are you sure that `$this->datetime` (optionally with the time added by a comma) is a correct format for the property?

File: system/modules/core/templates/elements/ce_accordion.html5 [hunk 3]
```diff
@@ -25,7 +25,7 @@
         <?php endif; ?>
 
         <?php if ($this->caption): ?>
-          <figcaption class="caption"><?php echo $this->caption; ?></figcaption>
+          <figcaption class="caption" itemprop="caption description"><?php echo $this->caption; ?></figcaption>
```
Comment (leofeyer) [inline system/modules/core/templates/elements/ce_accordion.html5]: I think it is wrong to use both `caption` and `description`. Isn't the description of an image the content of the `alt` attribute?

File: system/modules/core/templates/elements/ce_accordion.html5 [hunk 4]
```diff
@@ -12,10 +12,10 @@
     <?php endif; ?>
 
     <?php if ($this->addImage): ?>
-      <figure class="image_container<?php echo $this->floatClass; ?>"<?php if ($this->margin): ?> style="<?php echo $this->margin; ?>"<?php endif; ?>>
+      <figure class="image_container<?php echo $this->floatClass; ?>"<?php if ($this->margin): ?> style="<?php echo $this->margin; ?>"<?php endif; ?> itemprop="associatedMedia" itemscope itemtype="ImageObject">
```
Comment (leofeyer) [inline system/modules/core/templates/elements/ce_accordion.html5]: You have specified the other `itemtype` attributes as `http://schema.org/<type>`. Why did you omit the `http://schema.org/` here?

File: system/modules/core/templates/modules/mod_breadcrumb.html5 [hunk 5]
```diff
@@ -2,14 +2,14 @@
 
 <?php $this->block('content'); ?>
 
-  <ul>
+  <ol itemscope itemtype="http://schema.org/BreadcrumbList">
```
Comment (leofeyer) [inline system/modules/core/templates/modules/mod_breadcrumb.html5]: This is wrong (in terms of backwards compatibility). The list must remain unordered.

File: system/modules/core/templates/modules/mod_breadcrumb.html5 [hunk 6]
```diff
@@ -2,14 +2,14 @@
...
     <?php foreach ($this->items as $item): ?>
       <?php if ($item['isActive']): ?>
-        <li class="active<?php if ($item['class']): ?> <?php echo $item['class']; ?><?php endif; ?> last"><?php echo $item['title']; ?></li>
+        <li class="active<?php if ($item['class']): ?> <?php echo $item['class']; ?><?php endif; ?> last" itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem"><span itemprop="item name"><?php echo $item['title']; ?></span></li>
```
Comment (leofeyer) [inline system/modules/core/templates/modules/mod_breadcrumb.html5]: This is also a BC break, because you are adding a `<span>` element, which was not present before.

File: system/modules/core/templates/navigation/nav_default.html5 [hunk 7]
```diff
@@ -2,9 +2,9 @@
...
+      <li<?php if ($item['class']): ?> class="<?php echo $item['class']; ?>"<?php endif; ?>><a href="<?php echo $item['href']; ?>" title="<?php echo $item['pageTitle'] ?: $item['title']; ?>"<?php if ($item['class']): ?> class="<?php echo $item['class']; ?>"<?php endif; ?><?php if ($item['accesskey'] != ''): ?> accesskey="<?php echo $item['accesskey']; ?>"<?php endif; ?><?php if ($item['tabindex']): ?> tabindex="<?php echo $item['tabindex']; ?>"<?php endif; ?><?php if ($item['nofollow']): ?> rel="nofollow"<?php endif; ?><?php echo $item['target']; ?> role="menuitem"<?php if (!empty($item['subitems'])): ?> aria-haspopup="true"<?php endif; ?> itemprop="url"><span itemprop="name"><?php echo $item['link']; ?></span></a><?php echo $item['subitems']; ?></li>
```
Comment (leofeyer) [inline system/modules/core/templates/navigation/nav_default.html5]: Again, you have added a `<span>` attribute. And you have marked the URL as "name".

File: system/modules/core/templates/pagination/pagination.html5 [hunk 8]
```diff
@@ -22,7 +22,7 @@
     <?php endforeach; ?>
 
     <?php if ($this->hasNext): ?>
-      <li class="next"><a href="<?php echo $this->next['href']; ?>" class="next" title="<?php echo $this->next['title']; ?>"><?php echo $this->next['link']; ?></a></li>
+      <li class="next"><a href="<?php echo $this->next['href']; ?>" class="next" title="<?php echo $this->next['title']; ?>" rel="prev"><?php echo $this->next['link']; ?></a></li>
```
Comment (leofeyer) [inline system/modules/core/templates/pagination/pagination.html5]: This should be `rel="next"` I suppose.

File: system/modules/core/templates/picture/picture_default.html5 [hunk 9]
```diff
@@ -8,7 +8,9 @@
     <!--[if IE 9]></video><![endif]-->
 <?php endif; ?>
 
-<img src="<?php echo $this->img['src']; ?>"<?php if ($this->img['srcset'] !== $this->img['src']): ?> srcset="<?php echo $this->img['srcset']; ?>"<?php endif; ?><?php if (!empty($this->img['sizes'])): ?> sizes="<?php echo $this->img['sizes']; ?>"<?php elseif (!$this->sources): ?> width="<?php echo $this->img['width']; ?>" height="<?php echo $this->img['height']; ?>"<?php endif; ?> alt="<?php echo $this->alt; ?>"<?php if ($this->title): ?> title="<?php echo $this->title; ?>"<?php endif; ?>>
+<meta itemprop="width" content="<?php echo $this->img['width']; ?>">
```
Comment (leofeyer) [inline system/modules/core/templates/picture/picture_default.html5]: Why did you add two separate `<meta>` tags here?

Comment (marcobiedermann) [inline system/modules/core/templates/picture/picture_default.html5]: These are properties of the parent [ImageObject](http://schema.org/ImageObject) element

File: system/modules/core/templates/navigation/nav_default.html5 (hunk 7 above)
Comment (marcobiedermann) [inline system/modules/core/templates/navigation/nav_default.html5]: `itemprop="name"` needs to be wrapped inside a span or should be added to `<li>` tag

schema-li
schema-no-span
schema-span

File: system/modules/calendar/templates/calendar/cal_default.html5 (hunk 1 above)
Comment (marcobiedermann) [inline system/modules/calendar/templates/calendar/cal_default.html5]: `<span itemprop="name">` is inside the URL
The markup has to look like this
```
<div itemscope itemtype="http://schema.org/Event">
  <a href="#" itemprop="url">
    <span itemprop="name">Event Name</span>
    <time itemprop="startDate" content=""></time>
  </a>
</div>
```

I've added an extra span because `itemprop="name"` has to be a child of `itemtype="http://schema.org/Event"` and can not be added to the `<a>` tag. Otherwise the name would be the content of the anchor

File: system/modules/core/templates/pagination/pagination.html5 (hunk 8 above)
Comment (marcobiedermann) [inline system/modules/core/templates/pagination/pagination.html5]: Sorry, my bad!

File: system/modules/core/templates/modules/mod_breadcrumb.html5 (hunk 6 above)
Comment (marcobiedermann) [inline system/modules/core/templates/modules/mod_breadcrumb.html5]: `itemprop="name"` is a child of `itemtype="http://schema.org/ListItem"` so this has to be wrapped inside an extra span

File: system/modules/core/templates/picture/picture_default.html5 (hunk 9 above)
Comment (leofeyer) [inline system/modules/core/templates/picture/picture_default.html5]: The docs say

> This technique should be used sparingly. Only use meta with content for information that cannot otherwise be marked up.

Can we not mark it up differently?

File: system/modules/core/templates/modules/mod_breadcrumb.html5 [hunk 10]
```diff
@@ -2,12 +2,12 @@
 
 <?php $this->block('content'); ?>
 
-  <ul>
+  <ul itemscope itemtype="http://schema.org/BreadcrumbList">
     <?php foreach ($this->items as $item): ?>
       <?php if ($item['isActive']): ?>
-        <li class="active<?php if ($item['class']): ?> <?php echo $item['class']; ?><?php endif; ?> last"><?php echo $item['title']; ?></li>
+        <li class="active<?php if ($item['class']): ?> <?php echo $item['class']; ?><?php endif; ?> last" itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem"><span itemprop="item name"><?php echo $item['title']; ?></span></li>
```
Comment (leofeyer) [inline system/modules/core/templates/modules/mod_breadcrumb.html5]: Is `itemprop="item name"` correct? Or should it be `itemprop="name"`?

File: system/modules/core/templates/picture/picture_default.html5 (hunk 9 above)
Comment (marcobiedermann) [inline system/modules/core/templates/picture/picture_default.html5]: I think we got three solutions:
* omit it, because these values are present in `` tag
* add them via `<meta`> tag
* add them via `<span class="invisible">` tag

File: system/modules/core/templates/navigation/nav_default.html5 (hunk 7 above)
Comment (marcobiedermann) [inline system/modules/core/templates/navigation/nav_default.html5]: We actually have to add `itemprop="name"` to an inner `<span>` because otherwise this is not correct if there is a submenu inside a `<li>`

schema-navigation

File: system/modules/core/templates/picture/picture_default.html5 (hunk 9 above)
Comment (leofeyer) [inline system/modules/core/templates/picture/picture_default.html5]: I vote for omitting them.

**Comment by leofeyer** [comment]
I have begun to merge this PR in 3e956a773f2c4495761202b21e3e07f8dd6f2c4f.

**Comment by leofeyer** [comment]
According to the [docs](https://schema.org/docs/gs.html#advanced_dates), we can omit the `content` attribute of the `<time>` tag if there is an `datetime` attribute. @marcobiedermann Can you confirm?

**Comment by leofeyer** [comment]
The generated output of the calendar module seems to be invalid:

**Comment by leofeyer** [comment]
Außerdem fehlen bei mir die URLs im Seitenmenü:

**Comment by leofeyer** [comment]
Merged in 7801be5ebbf3d576db9ad7b125628ee28255598c.

**Comment by marcobiedermann** [comment]
Nice! Will check it out on deploy :)

**Comment by leofeyer** [comment]
I had to revert the change in 993e03bcf591bfddb3220a39331e7c218beae857. We'll ship the changes with Contao 4, because the template adjustments are not backwards compatible.

**Comment by marcobiedermann** [comment]
@leofeyer I've tried this markup for the calendar module
```
<td>
  <div itemscope itemtype="http://schema.org/Event">
    <a href="#" itemprop="url">
      1
      <span class="invisible">
        <span itemprop="name">Event Title</span>
        <span itemprop="location" itemscope itemtype="http://schema.org/Place">
          <span itemprop="name">Location</span>
          <address itemprop="address" itemscope itemtype="http://schema.org/PostalAddress">
            <span itemprop="name">Location</span>
          </address>
        </span>
        <time itemprop="startDate" content="2016-01-01T12:00:00+02:00"></time>
      </span>
    </a>
  </div>
</td>
```

Are you fine with this one? If so, I'll update my PR :)
Or do you wanna add optional fields for entering the full address (street, zip, city, country…)?

**Comment by leofeyer** [comment]
First, you cannot update the PR, because it has already been merged. If you want to add something, you will have to create a PR against the appropriate bundle (e.g. [contao/calendar-bundle](https://github.com/contao/calendar-bundle)).

Second, I'd prefer not to clutter the markup with so many additional tags, which don't serve an actual purpose. There is no way to enter a location address in Contao, therefore there is no way to output one, either. I guess we'll have to live with this.
